### DISCUSSION TITLE
Problem with adding a new key in an object state

### DISCUSSION THREAD

**Discussion by WalterRestifo** [neuefische/questions#445] [Web] [answered]
Hello everyone!

I get an object called Team1 (Team2 too, but for the sake of the example I will only talk about Team1) from local storage and save it in the state Team1.

I have an Input where I can save the score of the team. 
When I submit the form, the score get saved in the state pointsTeam1 and the state isSubmitted become true. I need that this state remain true until I route with a Link to another page. When I come back from another page, the state should be automatically reset to false, I think, because the whole page is loaded anew.

When isSubmitted change his state, the useEffect run, because of the dependency array. I set the team1 state to the old team1, but with a new key points, that has the value of the team1points state.
I save the new team1 state in the local storage under the name team1WithPoints, because I need this object in another page. 

The problem is that the whole logic works only when I use it the first time. If I test it a second time, team1WithPoints will not contain any key named points.

Why? (sorry, something doesn't work with adding code to the discussion. I tried to fix it) 

[PR](https://github.com/WalterRestifo/capstone-project-typescript/compare/feature/record-games?expand=1)

[US](https://github.com/WalterRestifo/capstone-project/issues/3)

This is my code: 
```jsx
export default function ScoreForm(): JSX.Element {
  const [isSubmitted, setIsSubmitted] = useState(false);
  const [pointsTeam1, setPointsTeam1] = useState(0);
  const [pointsTeam2, setPointsTeam2] = useState(0);
  const [team1, setTeam1] = useState<Team>({ players: [], points: 0 });
  const [team2, setTeam2] = useState<Team>({ players: [], points: 0 });

  useEffect(() => {
    if (typeof window !== "undefined") {
      const team1FromLocalStorage = localStorage.getItem("team1");
      if (team1FromLocalStorage) {
        setTeam1(JSON.parse(team1FromLocalStorage));
      }

      const team2FromLocalStorage = localStorage.getItem("team2");
      if (team2FromLocalStorage) {
        setTeam2(JSON.parse(team2FromLocalStorage));
      }
    }
  }, []);

 function handleSubmit(event: any) {
    event.preventDefault();
    const form = event.currentTarget;
    setPointsTeam1(form.elements.pointsTeam1.value);
    setPointsTeam2(form.elements.pointsTeam2.value);
    setIsSubmitted(true);
  }

  useEffect(() => {
    if (isSubmitted) {
      console.log("pointsTeam1: ", pointsTeam1);
      console.log("pointsTeam2: ", pointsTeam2);

      setTeam1({ ...team1, points: pointsTeam1 });
      localStorage.setItem("team1WithPoints", JSON.stringify(team1));
      console.log("team1: ", team1);

      setTeam2({ ...team2, points: pointsTeam2 });
      localStorage.setItem("team2WithPoints", JSON.stringify(team2));
      console.log("team2: ", team2);
    }
  }, [isSubmitted]);
 
```

**Comment by pixelass** [comment]
I updated your question with code blocks with the correct syntax highlighting.
Have a look at the suggested format for code blocks:

- https://github.com/neuefische/questions/wiki/How-to-ask-good-questions#react
- https://docs.github.com/en/get-started/writing-on-github/working-with-advanced-formatting/creating-and-highlighting-code-blocks

❤️‍🔥

**Reply by WalterRestifo** [reply]
Thank you!

**Comment by ThomasKirst** [comment, marked as answer]
You might have a problem in your logic, where you update the state of the team1 points.
Directly thereafter you attempt to store the presumably updated state into localStorage.
This won´t work as expected. From the beta.reactjs docs:
> The set function only updates the state variable for the next render. If you read the state variable after calling the set function, [you will still get the old value](https://beta.reactjs.org/reference/react/useState#ive-updated-the-state-but-logging-gives-me-the-old-value) that was on the screen before your call.

You can also pass a callback function into `setTeam1` and try something like this:
```js
setTeam1(oldState => {
  const newState = { ...team1, points: pointsTeam1 };
  localStorage.setItem("team1WithPoints", JSON.stringify(newState));
  return newState;
});
```
Instead of dealing with localStorage and state separately I would strongly recommend using a library like [use-local-storage-state](https://github.com/astoilkov/use-local-storage-state).

**Reply by WalterRestifo** [reply]
Thank you Thomas, you are right!
I wrote the code like this and it worked:
```

useEffect(() => {
    if (isSubmitted) {

      const team1WithPoints = { ...team1, points: pointsTeam1 };
      localStorage.setItem("team1WithPoints", JSON.stringify(team1WithPoints));

      const team2WithPoints = { ...team2, points: pointsTeam2 };
      localStorage.setItem("team2WithPoints", JSON.stringify(team2WithPoints));
    }
  }, [isSubmitted]);

```
